clang/lib/Basic/DiagnosticIDs.cpp

```diff
@@ -571,8 +571,7 @@ DiagnosticIDs::getDiagnosticSeverity(unsigned DiagID, SourceLocation Loc,
   }
 
   // If explicitly requested, map fatal errors to errors.
-  if (Result == diag::Severity::Fatal &&
-      Diag.CurDiagID != diag::fatal_too_many_errors && Diag.FatalsAsError)
```

Comment:
I'm not sure why do we checked for `Diag.CurDiagID != diag::fatal_too_many_errors` here.
Moreover even without this piece of code unit-tests succeed.

Reply:
That came from 878271b2945e89b868fb8daefea3dfaffce32710 and doesn't appear to have been tested. I think the logic is: if the severity is fatal, we would downgrade to error, but we do not want to downgrade the "too many fatal errors" diagnostic because that *has* to be fatal.

We could probably use better test coverage in this area; I suspect that code is still needed.

Reply:
I researched over the codebase and found many easy to prove usage patterns:
- `DiagID` comes from the same `Diag` object, e.g. https://github.com/llvm/llvm-project/blob/main/clang/lib/Basic/Diagnostic.cpp#L536
  => they're interchangeable
- code compares the result level with `> DiagnosticsEngine::Warning`, e.g. https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/Driver.cpp#L277
or `>= DiagnosticsEngine::Error` - https://github.com/llvm/llvm-project/blob/main/clang/lib/Sema/SemaDecl.cpp#L16351 
or `if (CurLevel < DiagnosticsEngine::Error)` - https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ASTReader.cpp#L498
  but both `Fatal` and `Error` are greater than `Warning` (or not less than `Error`)
  => these comparisons aren't affected at all

as well as some not so easy to prove:
- https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/ParsedAST.cpp#L578
  but it checks levels of a particular DiagIDs
  => probably we shouldn't rely on how many errors we faced before this checks (many of not so)
- https://github.com/llvm/llvm-project/blob/main/clang/tools/diagtool/ShowEnabledWarnings.cpp#L138
  ! here I'm not sure should we prevent downgrading some `Fatal`s to `Error`s here if the current diagnostic is in the `fatal_too_many_errors` "state"
- https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clang-tidy/ClangTidyDiagnosticConsumer.h#L217
  !! the hardest one to prove coz it has so many usages

Well, for now I would do it in a simple/hopeful way - just use the `DiagID` argument. But, if it doesn't seem safe to reviewers - I could add smth like `size_t count_in_flight__fatal_too_many_errors` to the `DiagnosticsEngine` to mimic the original behavior.

Reply:
> Well, for now I would do it in a simple/hopeful way - just use the DiagID argument. But, if it doesn't seem safe to reviewers - I could add smth like size_t count_in_flight__fatal_too_many_errors to the DiagnosticsEngine to mimic the original behavior.

I think it should be reasonably safe to use `DiagID`. We could add some test coverage near https://github.com/llvm/llvm-project/blob/2670565afc4ec855fa406b8f582dde44ce1739fb/clang/unittests/Basic/DiagnosticTest.cpp#L61 as it seems we can only downgrade fatal diagnostics to errors as part of internal APIs rather than user-facing flags. The test could look something like (I've not tried this out myself):
```
TEST(DiagnosticTest, fatalsAsError) {
  DiagnosticsEngine Diags(new DiagnosticIDs(),
                          new DiagnosticOptions,
                          new IgnoringDiagConsumer());
  Diags.setFatalsAsError(true);

  // Report a fatal_too_many_errors diagnostic to ensure that still
  // acts as a fatal error despite downgrading fatal errors to errors.
  Diags.Report(diag::fatal_too_many_errors);
  EXPECT_TRUE(Diags.hasFatalErrorOccurred());

  // Ensure that the severity of that diagnostic really is "fatal".
  EXPECT_EQ(Diags.getDiagnosticIDs()->getDiagnosticSeverity(diag::fatal_too_many_errors, {}, Diags), diags::Severity::Fatal);
}
```